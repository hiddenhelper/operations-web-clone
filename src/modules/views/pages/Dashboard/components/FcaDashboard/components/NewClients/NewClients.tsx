import React, { memo, useEffect, useMemo } from 'react';

import PieWidget from '../../../../../../shared/DashboardWidget/PieWidget';

import { ResourceModel, GeneralModel, StatisticsModel } from '../../../../../../../models';
import { parseQuery } from '../../../../../../../../utils/generalUtils';
import { ROUTES } from '../../../../../../../../constants';

export interface INewClientsProps {
  queryParams: GeneralModel.IQueryParams;
  pieStatisticsMap: GeneralModel.IEntityMap<StatisticsModel.IPieWidgetStatistics>;
  loading: GeneralModel.ILoadingStatus;
  fetchClientWidget: (key: string, query: GeneralModel.IQueryParams) => void;
}

const NewClients = ({ queryParams, pieStatisticsMap, loading, fetchClientWidget }: INewClientsProps) => {
  const idKey = useMemo(() => 'mainNewClients', []);
  const data = useMemo(() => pieStatisticsMap[idKey] || StatisticsModel.getPieWidgetStatisticsFallback(), [pieStatisticsMap, idKey]);

  const pieChartData = useMemo(
    () =>
      data.chart.map(entry => ({
        ...entry,
        x: ResourceModel.pieStatusMap[entry.status],
        y: entry.total,
        color: ResourceModel.statusColorMap[entry.status],
      })),
    [data]
  );

  const isClientWidgetLoading: boolean = useMemo(() => loading && loading.isLoading, [loading]);

  useEffect(() => {
    fetchClientWidget(idKey, queryParams);
  }, [queryParams, idKey, fetchClientWidget]);
  return (
    <PieWidget
      title="New Clients"
      total={data.newClients}
      data={pieChartData}
      isLoading={isClientWidgetLoading}
      detailRoute={`${ROUTES.DASHBOARD_CLIENTS.path}?${parseQuery(queryParams)}`}
    />
  );
};

export default memo(NewClients);
