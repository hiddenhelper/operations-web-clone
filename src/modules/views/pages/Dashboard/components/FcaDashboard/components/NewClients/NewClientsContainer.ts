import { Dispatch } from 'redux';
import { connect } from 'react-redux';

import { GeneralModel } from '../../../../../../../models';
import { GENERAL } from '../../../../../../../../constants';
import { IRootState } from '../../../../../../../state-mgmt/rootState';
import { statisticsState } from '../../../../../../../state-mgmt/statistics';

import NewClients from './NewClients';

export const mapStateToProps = (state: IRootState) => ({
  pieStatisticsMap: state.statistics.pieStatisticsMap,
  loading: state.general.loadingMap[GENERAL.LOADING_KEY.FETCH_CLIENT_WIDGET_STATISTICS],
});

export const mapDispatchToProps = (dispatch: Dispatch) => ({
  fetchClientWidget: (key: string, query: GeneralModel.IQueryParams) => dispatch(statisticsState.actions.fetchClientWidgetStatisticsStart(key, query)),
});

export default connect(mapStateToProps, mapDispatchToProps)(NewClients);
