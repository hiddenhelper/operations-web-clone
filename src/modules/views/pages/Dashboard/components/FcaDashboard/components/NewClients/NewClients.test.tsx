import React from 'react';
import { render, RenderResult } from '@testing-library/react';
import { MemoryRouter } from 'react-router';

import { getClientWidgetStatistics_1 } from '../../../../../../../../test/entities';
import NewClients, { INewClientsProps } from './NewClients';

describe('NewClients Component', () => {
  global.console.warn = () => {
    /** */
  };
  let wrapper: RenderResult;
  let props: INewClientsProps;

  beforeEach(() => {
    const mockDate = jest.fn(() => new Date(Date.UTC(2020, 11, 14)).valueOf());
    Date.now = mockDate;
    props = {
      queryParams: { stateCode: '', period: 0 },
      pieStatisticsMap: { mainNewClients: getClientWidgetStatistics_1() },
      loading: {
        isLoading: false,
        hasError: false,
        error: null,
      },
      fetchClientWidget: jest.fn(),
    };
  });

  it('should render', () => {
    wrapper = render(
      <MemoryRouter>
        <NewClients {...props} />
      </MemoryRouter>
    );
    expect(wrapper.container).toMatchSnapshot();
  });

  it('should fetchClientWidget', () => {
    props.pieStatisticsMap = {};
    render(
      <MemoryRouter>
        <NewClients {...props} />
      </MemoryRouter>
    );
    expect(props.fetchClientWidget).toHaveBeenCalled();
  });

  it('should render loading new clients', () => {
    props.loading = { isLoading: true, hasError: false, error: null };
    wrapper = render(
      <MemoryRouter>
        <NewClients {...props} />
      </MemoryRouter>
    );
    expect(wrapper.container).toMatchSnapshot();
  });
});
