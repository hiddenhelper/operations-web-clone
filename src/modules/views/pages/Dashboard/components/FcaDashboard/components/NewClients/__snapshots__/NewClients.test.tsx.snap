// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewClients Component should render 1`] = `
<div>
  <div
    class="MuiPaper-root makeStyles-maxHeight-1 makeStyles-pieContainer-2 makeStyles-widget-7 MuiPaper-elevation0 MuiPaper-rounded"
  >
    <div
      style="display: flex; justify-content: space-between;"
    >
      <div
        class="makeStyles-headerRow-37"
        style="width: 100%;"
      >
        <div
          class=""
        >
          <div
            class=" makeStyles-widgetInlineStatusWrapper-23"
          >
            <div
              class="makeStyles-widgetContentWrapper-14 makeStyles-widgetContentWrapperBottomMargin-15 makeStyles-marginNull-6"
            >
              <span
                class="makeStyles-widgetContent-19 makeStyles-widgetInlineMargin-21"
              >
                2
              </span>
              <p
                class="MuiTypography-root  makeStyles-widgetTitle-9 makeStyles-widgetTitleInlinePosition-24 MuiTypography-body1"
              >
                New Clients
              </p>
              <div
                class="makeStyles-chipStatusLeftMargin-29"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root makeStyles-pieWidgetDefault-41 MuiGrid-container"
    >
      <div
        class="MuiGrid-root makeStyles-pieWidgetItemDefault-42 MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-12 MuiGrid-grid-lg-12 MuiGrid-grid-xl-12"
      >
        <div
          class="MuiGrid-root MuiGrid-container"
          style="flex-direction: row; height: 100%;"
        >
          <div
            class="MuiGrid-root makeStyles-pieChartClass-46 MuiGrid-item MuiGrid-grid-lg-12 MuiGrid-grid-xl-12"
            style="padding-right: 0px; text-align: center; height: 100%;"
          >
            <div
              class="VictoryContainer"
              style="max-height: 275px; display: inline-block; width: auto; max-width: 255px; text-align: center; margin-bottom: 18px; height: 100%; pointer-events: none; position: relative;"
            >
              <svg
                height="400"
                role="img"
                style="pointer-events: all; width: 100%; height: 100%;"
                viewBox="0 0 400 400"
                width="400"
              >
                <g>
                  <path
                    d="M6.927352299505378,-149.8399539179005A150,150,0,0,1,6.927352299505378,149.8399539179005L6.927352299505364,129.81529875218305A130,130,0,0,0,6.927352299505364,-129.81529875218305Z"
                    role="presentation"
                    shape-rendering="auto"
                    style="fill: #37A66C; padding: 10px; stroke: transparent; stroke-width: 1;"
                    transform="translate(200, 200)"
                  />
                  <path
                    d="M-6.927352299505359,149.8399539179005A150,150,0,0,1,-6.927352299505329,-149.8399539179005L-6.9273522995053805,-129.81529875218305A130,130,0,0,0,-6.9273522995053485,129.81529875218305Z"
                    role="presentation"
                    shape-rendering="auto"
                    style="fill: #1F86E8; padding: 10px; stroke: transparent; stroke-width: 1;"
                    transform="translate(200, 200)"
                  />
                  <text
                    direction="inherit"
                    dx="0"
                    id="pie-labels-0"
                    x="370"
                    y="204.97"
                  >
                    <tspan
                      dx="0"
                      dy="0"
                      style="padding: 20px; font-family: 'Gill Sans', 'Seravek', 'Trebuchet MS', sans-serif; font-size: 14px; letter-spacing: normal; fill: #252525; stroke: transparent;"
                      text-anchor="start"
                      x="370"
                    />
                  </text>
                  <text
                    direction="inherit"
                    dx="0"
                    id="pie-labels-1"
                    x="30"
                    y="204.97"
                  >
                    <tspan
                      dx="0"
                      dy="0"
                      style="padding: 20px; font-family: 'Gill Sans', 'Seravek', 'Trebuchet MS', sans-serif; font-size: 14px; letter-spacing: normal; fill: #252525; stroke: transparent;"
                      text-anchor="end"
                      x="30"
                    />
                  </text>
                </g>
              </svg>
              <div
                style="z-index: 99; position: absolute; top: 0px; left: 0px; width: 100%; height: 100%;"
              >
                <svg
                  height="400"
                  style="overflow: visible; width: 100%; height: 100%;"
                  viewBox="0 0 400 400"
                  width="400"
                />
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-lg-12 MuiGrid-grid-xl-12"
          >
            <div
              class="MuiGrid-root MuiGrid-container"
            >
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-lg-9 MuiGrid-grid-xl-9"
              >
                <div
                  class=" makeStyles-bulletLabelWrapper-47"
                >
                  <span
                    class="makeStyles-statusCircle-48"
                    style="background-color: rgb(55, 166, 108);"
                  />
                  <p
                    class="MuiTypography-root makeStyles-statusText-49 MuiTypography-body1"
                  >
                    Active
                  </p>
                </div>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-lg-2 MuiGrid-grid-xl-2"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1"
                  style="color: rgb(76, 76, 76); font-size: 15px; text-align: right;"
                >
                  0.5
                  %
                </p>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-lg-1 MuiGrid-grid-xl-1"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1"
                  style="color: rgb(76, 76, 76); font-size: 15px; font-weight: bold; text-align: right;"
                >
                  1
                </p>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-container"
            >
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-lg-9 MuiGrid-grid-xl-9"
              >
                <div
                  class=" makeStyles-bulletLabelWrapper-47"
                >
                  <span
                    class="makeStyles-statusCircle-48"
                    style="background-color: rgb(233, 166, 31);"
                  />
                  <p
                    class="MuiTypography-root makeStyles-statusText-49 MuiTypography-body1"
                  >
                    Pending
                  </p>
                </div>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-lg-2 MuiGrid-grid-xl-2"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1"
                  style="color: rgb(76, 76, 76); font-size: 15px; text-align: right;"
                >
                  0
                  %
                </p>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-lg-1 MuiGrid-grid-xl-1"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1"
                  style="color: rgb(76, 76, 76); font-size: 15px; font-weight: bold; text-align: right;"
                >
                  0
                </p>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-container"
            >
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-lg-9 MuiGrid-grid-xl-9"
              >
                <div
                  class=" makeStyles-bulletLabelWrapper-47"
                >
                  <span
                    class="makeStyles-statusCircle-48"
                    style="background-color: rgb(31, 134, 232);"
                  />
                  <p
                    class="MuiTypography-root makeStyles-statusText-49 MuiTypography-body1"
                  >
                    Draft
                  </p>
                </div>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-lg-2 MuiGrid-grid-xl-2"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1"
                  style="color: rgb(76, 76, 76); font-size: 15px; text-align: right;"
                >
                  0.5
                  %
                </p>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-lg-1 MuiGrid-grid-xl-1"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1"
                  style="color: rgb(76, 76, 76); font-size: 15px; font-weight: bold; text-align: right;"
                >
                  1
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="makeStyles-widgetBulletLabelWrapper-3"
      style="display: flex; flex-direction: column;"
    >
      <hr
        class="MuiDivider-root makeStyles-widgetDivider-4 makeStyles-widgetDividerTopMargin-5"
      />
      <a
        class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
        href="/dashboard/clients?stateCode=&period=0"
      >
        <span
          class="makeStyles-widgetCta-28"
        >
          See more details
        </span>
      </a>
    </div>
  </div>
</div>
`;

exports[`NewClients Component should render loading new clients 1`] = `
<div>
  <div
    class="MuiPaper-root makeStyles-maxHeight-104 makeStyles-pieContainer-105 makeStyles-widget-110 MuiPaper-elevation0 MuiPaper-rounded"
  >
    <div
      style="display: flex; justify-content: space-between;"
    >
      <div
        class="MuiSkeleton-root MuiSkeleton-rect MuiSkeleton-wave"
        style="width: 100%; height: 30px;"
      />
    </div>
    <div
      class="MuiGrid-root makeStyles-pieWidgetDefault-144 MuiGrid-container"
    >
      <div
        class="MuiGrid-root makeStyles-pieWidgetItemDefault-145 MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-12 MuiGrid-grid-lg-12 MuiGrid-grid-xl-12"
      >
        <div
          class="MuiSkeleton-root MuiSkeleton-rect MuiSkeleton-wave"
          style="width: 100%; height: 96%;"
        />
      </div>
    </div>
    <div
      class="makeStyles-widgetBulletLabelWrapper-106"
      style="display: flex; flex-direction: column;"
    >
      <hr
        class="MuiDivider-root makeStyles-widgetDivider-107 makeStyles-widgetDividerTopMargin-108"
      />
      <div
        class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
        style="width: 120px; height: 20px;"
      />
    </div>
  </div>
</div>
`;
